lib/internal/fs/utils.js

```diff
@@ -448,29 +448,65 @@ function dateFromMs(ms) {
   return new Date(MathRound(Number(ms)));
 }
 
-function BigIntStats(dev, mode, nlink, uid, gid, rdev, blksize,
-                     ino, size, blocks,
-                     atimeNs, mtimeNs, ctimeNs, birthtimeNs) {
-  ReflectApply(StatsBase, this, [dev, mode, nlink, uid, gid, rdev, blksize,
-                                 ino, size, blocks]);
-
-  this.atimeMs = atimeNs / kNsPerMsBigInt;
-  this.mtimeMs = mtimeNs / kNsPerMsBigInt;
-  this.ctimeMs = ctimeNs / kNsPerMsBigInt;
-  this.birthtimeMs = birthtimeNs / kNsPerMsBigInt;
-  this.atimeNs = atimeNs;
-  this.mtimeNs = mtimeNs;
-  this.ctimeNs = ctimeNs;
-  this.birthtimeNs = birthtimeNs;
-  this.atime = dateFromMs(this.atimeMs);
-  this.mtime = dateFromMs(this.mtimeMs);
-  this.ctime = dateFromMs(this.ctimeMs);
-  this.birthtime = dateFromMs(this.birthtimeMs);
+class BigIntStats extends StatsBase {
+  #atime;
+  #mtime;
+  #ctime;
+  #birthtime;
+
+  constructor(dev, mode, nlink, uid, gid, rdev, blksize,
+              ino, size, blocks,
+              atimeNs, mtimeNs, ctimeNs, birthtimeNs) {
+    super(dev, mode, nlink, uid, gid, rdev, blksize,
+          ino, size, blocks);
+
+    this.atimeMs = atimeNs / kNsPerMsBigInt;
+    this.mtimeMs = mtimeNs / kNsPerMsBigInt;
+    this.ctimeMs = ctimeNs / kNsPerMsBigInt;
+    this.birthtimeMs = birthtimeNs / kNsPerMsBigInt;
+    this.atimeNs = atimeNs;
+    this.mtimeNs = mtimeNs;
+    this.ctimeNs = ctimeNs;
+    this.birthtimeNs = birthtimeNs;
+
+    ObjectDefineProperties(this, {
+      __proto__: null,
+      atime: {
+        __proto__: null,
+        enumerable: true,
+        get() {
+          this.#atime ??= dateFromMs(this.atimeMs);
+          return this.#atime;
+        },
+      },
```

Comment:
I think it will be faster to have `get atime` instead.

Reply:
That breaks the tests and makes this PR a semver-major

Reply:
Hm... in this case, at least cache this object outside the constructor.

And without `set`, this is already a `semver-major`.

Reply:
The property descriptors could be stored on the class as static privates then be used with define properties like you have. I think this would improve perf a bit because each instance will no longer need to create it's own getter functions and objects for define properties, and V8 should be able to optimize the getters across all function instances. I can test that theory tomorrow.

Reply:
Shouldn't these getters/setters be defined on the prototype of the class rather than for each instance? I wouldn't be surprised if defining them for each instance is more expensive than creating the 4 Date objects in the first place.

Reply:
Updated the benchmark to include more cases, and the ones that keep backward compat unfortunately perform worse than eagerly creating the dates. https://github.com/nodejs/node/pull/50908#issuecomment-1828445121